cartridge/failover/raft.lua

```diff
@@ -0,0 +1,137 @@
+            and member.payload.leader_uuid ~= nil
+            and member.payload.raft_term >= last_term
+            then
+                last_leader = member.payload.leader_uuid
+                last_term = member.payload.raft_term
+            end
+        end
+        appointments[replicaset_uuid] = last_leader
+        ::next_rs::
+    end
+
+    appointments[vars.replicaset_uuid] = vars.leader_uuid
+    return appointments
+end
+
+local function on_election_trigger()
+    local election = box.info.election
+
+    local leader = box.info.replication[election.leader] or {}
+
+    if vars.leader_uuid ~= leader.uuid then
+        vars.cache.is_leader = vars.leader_uuid == vars.instance_uuid
+        vars.leader_uuid = leader.uuid
+        membership.set_payload('leader_uuid', vars.leader_uuid)
+    end
+    membership.set_payload('raft_term', election.term)
+end
+
+local function cfg()
+    assert(box.ctl.on_election, "Your Tarantool version doesn't support raft failover mode")
+    log.warn('Raft failover is in beta-version')
+    local box_opts = argparse.get_box_opts()
+
+    box.cfg{
+        -- The instance is set to candidate, so it may become leader itself
+        -- as well as vote for other instances.
+        --
+        -- Alternative: set one of instances to `voter`, so that it
+        -- never becomes a leader but still votes for one of its peers and helps
+        -- it reach election quorum.
+        election_mode = box_opts.election_mode or 'candidate',
+        -- Quorum for both synchronous transactions and
+        -- leader election votes.
+        replication_synchro_quorum = box_opts.replication_synchro_quorum or 'N/2 + 1',
+        -- Synchronous replication timeout. The transaction will be
+        -- rolled back if no quorum is achieved during timeout.
+        replication_synchro_timeout = box_opts.replication_synchro_timeout,
+        -- Heartbeat timeout. A leader is considered dead if it doesn't
+        -- send heartbeats for 4 * replication_timeout.
+        -- Once the leader is dead, remaining instances start a new election round.
+        replication_timeout = box_opts.replication_timeout,
+        -- Timeout between elections. Needed to restart elections when no leader
+        -- emerges soon enough. Equals 4 * replication_timeout
+        election_timeout = box_opts.election_timeout,
+    }
+
+    if vars.raft_trigger == nil then
+        vars.raft_trigger = box.ctl.on_election(on_election_trigger)
+    end
+
+    membership.set_payload('raft_term', box.info.election.term)
+
+    return true
+end
+
+-- disable raft if it was enabled
+local function disable()
+    if vars.raft_trigger ~= nil then
+        box.ctl.on_election(nil, vars.raft_trigger)
+        vars.raft_trigger = nil
+    end
+    box.cfg{ election_mode = 'off' }
+    vars.leader_uuid = nil
+end
+
+local function promote(replicaset_leaders)
+    local servers_list = package.loaded.cartridge.confapplier.get_readonly('topology').servers
```

Comment:
I'm not sure but in case of advertise_uri change real value could be different. And "refine_uri..." function should be used.